Source/ObjectiveDropboxOfficial/Shared/Handwritten/OAuth/DBAccessTokenProviderImpl.m

```diff
@@ -0,0 +1,116 @@
+///
+/// Copyright (c) 2020 Dropbox, Inc. All rights reserved.
+///
+
+#import "DBAccessTokenProvider+Internal.h"
+
+#import "DBOAuthResult.h"
+
+@implementation DBLongLivedAccessTokenProvider
+
+@synthesize accessToken = _accessToken;
+
+- (instancetype)initWithTokenString:(NSString *)tokenString {
+  self = [super init];
+  if (self) {
+    _accessToken = tokenString;
+  }
+  return self;
+}
+
+- (void)refreshAccessTokenIfNecessary:(DBOAuthCompletion)completion {
+  // Complete with empty result, because it doesn't need a refresh.
+  completion(nil);
+}
+
+@end
+
+@interface DBShortLivedAccessTokenProvider ()
+
+@property (nonatomic, strong) DBAccessToken *token;
+@property (nonatomic, strong) id<DBAccessTokenRefreshing> tokenRefresher;
+@property (nonatomic, strong) dispatch_queue_t queue;
+@property (nonatomic, strong) NSMutableArray<DBOAuthCompletion> *completionBlocks;
+
+
+@end
+
+@implementation DBShortLivedAccessTokenProvider
+
+- (instancetype)initWithToken:(DBAccessToken *)token tokenRefresher:(id<DBAccessTokenRefreshing>)tokenRefresher {
+  self = [super init];
+  if (self) {
+    _token = token;
+    _tokenRefresher = tokenRefresher;
+    dispatch_queue_attr_t qosAttribute =
+      dispatch_queue_attr_make_with_qos_class(DISPATCH_QUEUE_CONCURRENT, QOS_CLASS_USER_INITIATED, 0);
+    _queue = dispatch_queue_create("com.dropbox.dropbox_sdk_obj_c.DBShortLivedAccessTokenProvider.queue", qosAttribute);
+    _completionBlocks = [NSMutableArray new];
+  }
+  return self;
+}
+
+- (NSString *)accessToken {
+  __block NSString *tokenString = nil;
+  dispatch_sync(_queue, ^{
+    tokenString = _token.accessToken;
+  });
+  return tokenString;
+}
+
+- (void)refreshAccessTokenIfNecessary:(DBOAuthCompletion)completion {
+  dispatch_barrier_async(_queue, ^{
+    if (![self db_shouldRefresh]) {
+      completion(nil);
+      return;
+    }
+
+    // Ensure subsequent calls don't initiate more refresh requests, if one is in progress.
+    BOOL refreshInProgress = [self db_refreshInProgress];
+    [self->_completionBlocks addObject:completion];
+    if (!refreshInProgress) {
+      __weak typeof(self) weakSelf = self;
+      DBOAuthCompletion refreshCompletionBlock = ^(DBOAuthResult *result) {
+        typeof(self) strongSelf = weakSelf;
+        if (strongSelf) {
+          dispatch_async(strongSelf->_queue, ^{
```

Comment:
I think this has to be on the barrier. See outdated convo. 

> Right, it dispatches back on the queue passed in. 

> https://github.com/dropbox/dropbox-sdk-obj-c/pull/264/files#diff-7a40580386a62f3cc30c123bb74b44cdR101

> I think the problem here is not that. Its that the fact that the dispatch is not happening on the barrier. 

> _token.accessToken is assessed on a concurrent queue. It must be safely written by dispatching on barrier.

Reply:
oops, you are correct, all writes need to happen with barrier. thanks for catching this.